File: java/app/src/main/java/lox/Scanner.java
package lox;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static lox.TokenType.*;

public class Scanner {
	private static final Map<String, TokenType> keywords;

	static {
		keywords = new HashMap<>();
		keywords.put("and", AND);
		keywords.put("class", CLASS);
		keywords.put("else", ELSE);
		keywords.put("false", FALSE);
		keywords.put("for", FOR);
		keywords.put("fun", FUN);
		keywords.put("if", IF);
		keywords.put("nil", NIL);
		keywords.put("or", OR);
		keywords.put("print", PRINT);
		keywords.put("return", RETURN);
		keywords.put("super", SUPER);
		keywords.put("this", THIS);
		keywords.put("true", TRUE);
		keywords.put("var", VAR);
		keywords.put("while", WHILE);
	}

	private final String source;
	private final List<Token> tokens = new ArrayList<>();

	private int start = 0;
	private int current = 0;
	private int line = 1;

	public Scanner(String source) {
		this.source = source;
	}

	public List<Token> scanTokens() {
		while (!this.isAtEnd()) {
			// we are at the beginning of the next lexeme
			this.start = this.current;
			this.scanToken();
		}

		this.tokens.add(new Token(EOF, "", null, this.line));
		return this.tokens;
	}

	private void addToken(TokenType type) {
		this.addToken(type, null);
	}

	private void addToken(TokenType type, Object literal) {
		String text = this.source.substring(this.start, this.current);
		this.tokens.add(new Token(type, text, literal, this.line));
	}

	private char advance() {
		return this.source.charAt(this.current++);
	}

	private void blockComment() {
		for (;;) {
			char c = this.advance();
			switch (c) {
				case '\n':
					this.line++;
					break;

				case '/':
					if (this.match('*')) {
						// recurse into block comment
						this.blockComment();
					}
					break;

				case '*':
					if (this.match('/')) {
						return;
					}

				default:
					break;
			}
		}
	}

	private void identifier() {
		while (isAlphaNumeric(this.peek()))
			this.advance();

		var text = this.source.substring(this.start, this.current);
		var type = keywords.get(text);
		if (type == null) {
			type = IDENTIFIER;
		}

		this.addToken(type);
	}

	private boolean isAtEnd() {
		return this.current >= this.source.length();
	}

	private boolean match(char expected) {
		if (this.isAtEnd())
			return false;
		if (this.source.charAt(this.current) != expected)
			return false;

		this.current++;
		return true;
	}

	private void number() {
		while (isDigit(this.peek()))
			this.advance();

		// look for a fractional part
		if (this.peek() == '.' && isDigit(this.peekNext())) {
			// consume the "."
			this.advance();

			while (isDigit(this.peek()))
				advance();
		}

		this.addToken(NUMBER, Double.parseDouble(this.source.substring(this.start, this.current)));
	}

	private char peek() {
		if (this.isAtEnd())
			return '\0';
		return this.source.charAt(this.current);
	}

	private char peekNext() {
		var next = this.current + 1;
		if (next >= this.source.length())
			return '\0';
		return this.source.charAt(next);
	}

	private void scanToken() {
		char c = this.advance();
		switch (c) {
			case '(':
				this.addToken(LEFT_PAREN);
				break;
			case ')':
				this.addToken(RIGHT_PAREN);
				break;
			case '{':
				this.addToken(LEFT_BRACE);
				break;
			case '}':
				this.addToken(RIGHT_BRACE);
				break;
			case ',':
				this.addToken(COMMA);
				break;
			case '.':
				this.addToken(DOT);
				break;
			case '-':
				this.addToken(MINUS);
				break;
			case '+':
				this.addToken(PLUS);
				break;
			case ';':
				this.addToken(SEMICOLON);
				break;
			case '*':
				this.addToken(STAR);
				break;
			case '!':
				this.addToken(this.match('=') ? BANG_EQUAL : BANG);
				break;
			case '=':
				this.addToken(this.match('=') ? EQUAL_EQUAL : EQUAL);
				break;
			case '<':
				this.addToken(this.match('=') ? LESS_EQUAL : LESS);
				break;
			case '>':
				this.addToken(this.match('=') ? GREATER_EQUAL : GREATER);
				break;
			case '/':
				if (this.match('/')) {
					// a comment goes until the end of a line
					while (this.peek() != '\n' && !this.isAtEnd())
						this.advance();
				} else if (this.match('*')) {
					this.blockComment();
				} else {
					this.addToken(SLASH);
				}
				break;

			case '"':
				this.string();
				break;

			case '\n':
				this.line++;
			case ' ':
			case '\r':
			case '\t':
				// ignore whitespace
				break;

			default:
				if (isDigit(c)) {
					this.number();
				} else if (isAlpha(c)) {
					this.identifier();
				} else {
					Lox.error(this.line, "Unexpected character.");
				}
				break;
		}
	}

	private void string() {
		while (this.peek() != '"' && !this.isAtEnd()) {
			if (this.peek() == '\n')
				this.line++;
			this.advance();
		}

		if (this.isAtEnd()) {
			Lox.error(this.line, "Unterminated string.");
			return;
		}

		// The closing ".
		this.advance();

		// Trim surrounding ""
		var string = source.substring(this.start + 1, this.current - 1);
		this.addToken(STRING, string);
	}

	private static boolean isAlpha(char c) {
		return (c >= 'a' && c <= 'z') ||
				(c >= 'A' && c <= 'Z') ||
				c == '_';
	}

	private static boolean isAlphaNumeric(char c) {
		return isAlpha(c) || isDigit(c);
	}

	private static boolean isDigit(char c) {
		return c >= '0' && c <= '9';
	}
}

File: rust/src/scanner.rs
mod token;
mod token_kind;

pub use token::*;
pub use token_kind::*;

pub struct Scanner<'source> {
	start:   &'source str,
	current: &'source str,
	line:    u32,
}

impl<'source> Scanner<'source> {
	pub fn new(source: &'source str) -> Self {
		Self {
			start:   source,
			current: source,
			line:    1,
		}
	}

	pub fn scan_token(&mut self) -> Token {
		self.skip_whitespace();
		self.start = self.current;

		if self.is_at_end() {
			return self.make_token(TokenKind::Eof);
		}

		let Some(ch) = self.advance() else {
			return self.error_token("Unexpected character.");
		};

		match ch {
			ch if is_alpha(ch) => self.identifier(),
			ch if is_digit(ch) => self.number(),
			'"' => self.string(),
			'(' => self.make_token(TokenKind::LParen),
			')' => self.make_token(TokenKind::RParen),
			'{' => self.make_token(TokenKind::LBrace),
			'}' => self.make_token(TokenKind::RBrace),
			';' => self.make_token(TokenKind::Semicolon),
			',' => self.make_token(TokenKind::Comma),
			'.' => self.make_token(TokenKind::Dot),
			'-' => self.make_token(TokenKind::Minus),
			'+' => self.make_token(TokenKind::Plus),
			'/' => self.make_token(TokenKind::Slash),
			'*' => self.make_token(TokenKind::Star),

			'!' => self.consume_match('=', |scanner| {
				scanner.make_token(TokenKind::BangEqual)
			}).unwrap_or_else(|| {
				self.make_token(TokenKind::Bang)
			}),
			'=' => self.consume_match('=', |scanner| {
				scanner.make_token(TokenKind::EqualEqual)
			}).unwrap_or_else(|| {
				self.make_token(TokenKind::Equal)
			}),
			'<' => self.consume_match('=', |scanner| {
				scanner.make_token(TokenKind::LessEqual)
			}).unwrap_or_else(|| {
				self.make_token(TokenKind::Less)
			}),
			'>' => self.consume_match('=', |scanner| {
				scanner.make_token(TokenKind::GreaterEqual)
			}).unwrap_or_else(|| {
				self.make_token(TokenKind::Greater)
			}),

			_ => self.error_token("Unexpected character."),
		}
	}
}

impl<'source> Scanner<'source> {
	fn error_token(&self, msg: &str) -> Token {
		todo!()
	}

	fn make_token(&self, kind: TokenKind) -> Token<'source> {
		todo!()
	}
}

impl<'source> Scanner<'source> {
	fn advance(&mut self) -> Option<char> {
		self.peek().map(|ch| {
			self.current = &self.current[1..];
			ch
		})
	}

	fn identifier(&mut self) -> Token {
		todo!()
	}

	fn is_at_end(&self) -> bool {
		self.current.len() == 0
	}

	fn consume_match<F, T>(&mut self, expect: char, eval: F) -> Option<T>
	where
		F: FnOnce(&mut Self) -> T,
	{
		todo!()
	}

	fn number(&mut self) -> Token {
		todo!()
	}

	fn peek(&self) -> Option<char> {
		self.current.chars().next()
	}

	fn peek_next(&self) -> Option<char> {
		self.current.chars().skip(1).next()
	}

	fn skip_whitespace(&mut self) {
		loop {
			let Some(ch) = self.peek() else {
				return;
			};

			match ch {
				' ' | '\r' | '\t' => _ = self.advance(),
				'\n' => {
					self.line += 1;
					self.advance();
				},

				'/' => {
					if let Some('/') = self.peek_next() {
						while self.peek() != Some('\n') && !self.is_at_end() {
							self.advance();
						}
					} else {
						return;
					}
				},
				_ => return,
			}
		}
	}

	fn string(&mut self) -> Token {
		todo!()
	}
}

fn is_alpha(ch: char) -> bool {
	matches!(ch, 'a'..='z' | 'A'..='Z' | '_')
}

fn is_digit(ch: char) -> bool {
	matches!(ch, '0'..='9')
}
